base/applications/cmdutils/where/where.c

```diff
@@ -0,0 +1,564 @@
+    INT cch, iExt;
+    HANDLE hFind;
+    WIN32_FIND_DATAW find;
+    BOOL ret = TRUE;
+
+    for (iExt = 0; iExt < s_pathext.count; ++iExt)
+    {
+        // build path
+        StringCchCopyW(szPath, _countof(szPath), dir);
+        StringCchCatW(szPath, _countof(szPath), L"\\");
+        StringCchCatW(szPath, _countof(szPath), filename);
+        StringCchCatW(szPath, _countof(szPath), strlist_get_at(&s_pathext, iExt));
+
+        // enumerate file items
+        hFind = FindFirstFileExW(szPath, FindExInfoStandard, &find, FindExSearchNameMatch,
+                                 NULL, 0);
+        if (hFind != INVALID_HANDLE_VALUE)
+        {
+            pch = wcsrchr(szPath, L'\\') + 1; // file title
+            cch = _countof(szPath) - (pch - szPath); // remainder
+            do
+            {
+                if (find.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
+                    continue; // it was directory
+
+                StringCchCopyW(pch, cch, find.cFileName); // build full path
+
+                if (!WherePrintPath(szPath))
+                {
+                    WhereError(IDS_OUTOFMEMORY);
+                    ret = FALSE;
+                    break;
+                }
+            } while (FindNextFile(hFind, &find));
+            FindClose(hFind);
+        }
+        if (!ret)
+            break;
+    }
+
+    return ret;
+}
+
+// FIXME: Too slow. Optimize for speed.
+static BOOL WhereSearchRecursive(LPCWSTR filename, LPCWSTR dir)
+{
+    WCHAR szPath[MAX_PATH];
+    LPWSTR pch;
+    INT cch;
+    HANDLE hFind;
+    WIN32_FIND_DATAW find;
+    BOOL ret = WhereSearchFiles(filename, dir); // search files in the directory
+    if (!ret)
+        return ret;
+
+    // build path with wildcard
+    StringCchCopyW(szPath, _countof(szPath), dir);
+    StringCchCatW(szPath, _countof(szPath), L"\\*");
+
+    // enumerate directory items
+    hFind = FindFirstFileExW(szPath, FindExInfoStandard, &find, FindExSearchNameMatch,
+                             NULL, 0);
+    if (hFind != INVALID_HANDLE_VALUE)
+    {
+        pch = wcsrchr(szPath, L'\\') + 1; // file title
+        cch = _countof(szPath) - (pch - szPath); // remainder
+        do
+        {
+            if (!(find.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY))
+                continue; // not directory
+
+            if (IS_DOTS(find.cFileName))
+                continue; // ignore "." and ".."
+
+            StringCchCopyW(pch, cch, find.cFileName); // build full path
+
+            ret = WhereSearchRecursive(filename, szPath); // recurse
+            if (!ret)
+                break;
+        } while (FindNextFileW(hFind, &find));
+        FindClose(hFind);
+    }
+
+    return ret;
+}
+
+static BOOL WhereSearch(LPCWSTR filename, strlist_t *dirlist)
+{
+    INT iDir;
+    for (iDir = 0; iDir < dirlist->count; ++iDir) // for each directory
+    {
+        LPWSTR dir = strlist_get_at(dirlist, iDir);
+        BOOL ret = WhereSearchFiles(filename, dir); // search
```

Comment:
It's nice to have comments in the code, but when they paraphrase 1-to-1 the name of the functions, then they serve no purpose at all and are better be removed.
Comments are to explain the idea behind the code.